airbyte-integrations/connectors/destination-s3/src/main/java/io/airbyte/integrations/destination/s3/avro/JsonToAvroSchemaConverter.java

```diff
@@ -200,17 +200,22 @@ Schema getSingleFieldType(final String fieldName, final JsonSchemaType fieldType
     return fieldSchema;
   }
 
-  List<Schema> getSchemasFromTypes(final String fieldName, final ArrayNode types) {
-    return MoreIterators.toList(types.elements())
-        .stream()
-        .flatMap(definition -> getNonNullTypes(fieldName, definition).stream().flatMap(type -> {
-          final Schema singleFieldSchema = getSingleFieldType(fieldName, type, definition);
-          if (singleFieldSchema.isUnion()) {
-            return singleFieldSchema.getTypes().stream();
-          } else {
-            return Stream.of(singleFieldSchema);
-          }
-        }))
+  List<Schema> getSchemasFromTypes(final String fieldName, final ArrayNode types, boolean combined) {
+    return StreamUtils
+        .zipWithIndex(MoreIterators.toList(types.elements())
+            .stream())
+        .flatMap(definition -> getNonNullTypes(fieldName, definition.getValue()).stream()
+            .flatMap(type -> {
+              final Schema singleFieldSchema = getSingleFieldType(combined
+                  // avro schema allows only unique names of records in case of union
+                  ? fieldName.concat("_").concat(String.valueOf(definition.getIndex()))
+                  : fieldName, type, definition.getValue());
```

Comment:
We can extract to separate method getCombinedFieldName(combined, fieldName, definition.getIndex()) for better reading